aggregators/converter.go

```diff
@@ -25,114 +30,123 @@ const (
 	overflowBucketName = "_other"
 )
 
-func setMetricCountBasedOnOutcome(stm *ServiceTransactionMetrics, from *modelpb.APMEvent) {
-	txn := from.GetTransaction()
-	switch from.GetEvent().GetOutcome() {
-	case "failure":
-		stm.FailureCount = txn.GetRepresentativeCount()
-	case "success":
-		stm.SuccessCount = txn.GetRepresentativeCount()
-	}
-}
-
-// EventToCombinedMetrics converts APMEvent to one or more CombinedMetrics.
-// If an event results in multiple metrics, they may be spread across different partitions.
+// EventToCombinedMetrics converts APMEvent to one or more CombinedMetrics and
+// calls the provided callback for each pair of CombinedMetricsKey and
+// CombinedMetrics. The callback MUST NOT hold the reference of the passed
+// CombinedMetrics. If required, the callback can call CloneVT to clone the
+// CombinedMetrics. If an event results in multiple metrics, they may be spread
+// across different partitions.
 //
-// EventToCombinedMetrics will never produce overflow metrics, as it applies to a single APMEvent.
+// EventToCombinedMetrics will never produce overflow metrics, as it applies to a
+// single APMEvent.
 func EventToCombinedMetrics(
 	e *modelpb.APMEvent,
 	unpartitionedKey CombinedMetricsKey,
 	partitioner Partitioner,
-) (map[CombinedMetricsKey]*CombinedMetrics, error) {
-	var gl GlobalLabels
-	gl.fromLabelsAndNumericLabels(e.Labels, e.NumericLabels)
-	gls, err := gl.MarshalString()
+	callback func(CombinedMetricsKey, *aggregationpb.CombinedMetrics) error,
+) error {
+	svcKey := serviceKey(e, unpartitionedKey.Interval)
+	svcInstanceKey, err := serviceInstanceKey(e)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	hasher := Hasher{}.
+		Chain(serviceKeyHasher(svcKey)).
+		Chain(serviceInstanceKeyHasher(svcInstanceKey))
 
-	kvs := make(map[CombinedMetricsKey]*CombinedMetrics)
-	svcKey := serviceKey(e, unpartitionedKey.Interval)
-	svcInstanceKey := ServiceInstanceAggregationKey{GlobalLabelsStr: gls}
-	hasher := Hasher{}.Chain(svcKey).Chain(svcInstanceKey)
-
-	setCombinedMetrics := func(k CombinedMetricsKey, sim ServiceInstanceMetrics) {
-		cm, ok := kvs[k]
+	// m collects service instance metrics for each partition
```

Comment:
Have you looked at writing every metric as a separate merge operation, and deferring all merging to pebble? I don't like that we have merging in two places. I think this should be possible by first calculating how many metrics we may write, to figure out how to distribute the event count, and then assigning it to each record as we go. But I don't know about the performance difference.

Reply:
:ok_hand: maybe we can come back to it when you have removed translation from merging